File: slides/01-mathematical-concepts/rsrc/taylor_univariate.py
import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(-15, 15, 1000)
y = np.zeros(len(x))

a = 2

plt.figure(figsize=(10, 8))

plt.plot(x, np.sin(x), 'k', label='sin(x)')

y0 = y + np.sin(a)
plt.plot(x, y0, color='red', label='m=0')

y1 = y0 + np.cos(a) * (x-a)/np.math.factorial(1)
plt.plot(x, y1, color='blue', label='m=1')

y2 = y1 - np.sin(a) * ((x-a)**2)/np.math.factorial(2)
plt.plot(x, y2, color='orange', label='m=2')

y3 = y2 - np.cos(a) * ((x-a)**3)/np.math.factorial(3)
plt.plot(x, y3, color='green', label='m=3')

y4 = y3 + np.sin(a) * ((x-a)**4)/np.math.factorial(4)
plt.plot(x, y4, color='cyan', label='m=4')

plt.grid()
plt.title('Taylor polynomial for various orders at a=2', fontsize=20)
plt.xlabel('x', fontsize=20)
plt.ylabel('y', fontsize=20)
plt.xlim(-6, 10)
plt.ylim(-8, 8)
plt.legend(prop={'size': 15})
plt.savefig('C:/Users/mariu/Documents/Arbeit/SLDS/Optimization/1.2/taylor_univariate.png')
plt.show()

File: slides/01-mathematical-concepts/rsrc/taylor_univariate.R
# ------------------------------------------------------------------------------
# mathematical concepts

# FIG: plot taylor polynomial of 0-4 orders for sin(x).
# ------------------------------------------------------------------------------

library(ggplot2)
library(dplyr)

# DATA -------------------------------------------------------------------------

x <- seq(-15, 15, length.out = 1000)
a <- 2  # Expansion point

# Taylor series approximations
y0 <- rep(sin(a), length(x))
y1 <- y0 + cos(a) * (x - a) / factorial(1)
y2 <- y1 - sin(a) * (x - a)^2 / factorial(2)
y3 <- y2 - cos(a) * (x - a)^3 / factorial(3)
y4 <- y3 + sin(a) * (x - a)^4 / factorial(4)

df <- data.frame(
  x = rep(x, 6),
  y = c(sin(x), y0, y1, y2, y3, y4),
  order = factor(rep(c("sin(x)", "m=0", "m=1", "m=2", "m=3", "m=4"), each = length(x)),
                 levels = c("sin(x)", "m=0", "m=1", "m=2", "m=3", "m=4"))
)

# PLOT -------------------------------------------------------------------------

colors <- c("black", "red", "blue", "orange", "green", "cyan")

plot <- ggplot(df, aes(x = x, y = y, color = order)) +
  geom_line(size = 1) +
  scale_color_manual(values = colors) +
  labs(
    title = "Taylor polynomial for various orders at a=2",
    x = "x",
    y = "y",
    color = "Order"
  ) +
  xlim(-6, 10) + ylim(-8, 8) +
  theme_minimal() +
  theme(
    text = element_text(size = 20),
    legend.position = "right",
    legend.title = element_blank(),
    legend.text = element_text(size = 15)
  )

plot
ggsave("../figure_man/taylor_univariate.png", plot, width = 10, height = 8)
